File: LeetCode/LeetCode-CN-JAVA/2020六月每日一题/15. 三数之和.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> ans = new ArrayList<List<Integer>>();
        Arrays.sort(nums);
        int n = nums.length;
        // 确定a之后，查找所有b和c的组合
        for(int a = 0; a < n && nums[a] <= 0; a++) {
            if(a > 0 && nums[a] == nums[a - 1]) { // 如果a已经查找过了，则直接跳过
                continue;
            }
            int c = n - 1;
            // 确定a和b，查找所有符合条件的c
            for(int b = a + 1; b < n && nums[a] + nums[b] <= 0 ; b++) {
                if(b > a + 1 && nums[b] == nums[b - 1]) { // 如果b查找过则不再查找
                    continue;
                }
                while(b < c && nums[a] + nums[b] + nums[c] > 0) {
                    c--;
                }
                // b==c说明b后面的所有c，都是a+b+c>0，那么随着b和c的增加，则不会出现a+b+c==0的情况
                if(b == c) {
                    break;
                }
                // <0不处理，因为可能随着后边b的增加满足等于
                if(nums[a] + nums[b] + nums[c] == 0) {
                    ArrayList<Integer> subans = new ArrayList<>();
                    subans.add(nums[a]);
                    subans.add(nums[b]);
                    subans.add(nums[c]);
                    ans.add(subans);
                }
            }
        }
        return ans;
    }
}

File: LeetCode/LeetCode-CN-CPP/2020六月每日一题/15. 三数之和.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ans;
        sort(nums.begin(), nums.end()); // 首先进行排序
        for(int a = 0; a < nums.size(); a++) { // 先确定一个a，然后找出在a确定下的所有b、c组合
            if(a > 0 && nums[a] == nums[a - 1]) { // 如果当前的a已经查找过了，跳过
                continue;
            }
            int c = nums.size() - 1; // a+b+c=0时，当a确定了，b随着循环增大，如果要满足公式，则c随着循环变小，所以c从后面还是遍历
            for(int b = a + 1; b < nums.size(); b++) { // 再确定一个b , 此时确定了a和b，然后找到所有符合条件的c
                if(b > a + 1 && nums[b] == nums[b - 1]) { // 如果当前的a、b已经查找过了，跳过
                    continue;
                }
                // c的位置不可能比b小，如果此时a+b+c>0则说明c过大，需要小一点
                while(b < c && nums[a] + nums[b] + nums[c] > 0) {
                    c--;
                }
                // 跳出上面while可能是b==c or a+b+c <= 0
                // 如果是b==c ，说明b后面的所有的c都是a+b+c>0，并且随着b的增大，后续所有的式子都会>0
                if(b == c) {
                    break;
                }
                // 满足条件。<0时不用处理，因为随着b增大可能就满足了
                if(nums[a] + nums[b] + nums[c] == 0) {
                    ans.push_back({nums[a],nums[b],nums[c]});
                }
            }
        }
        return ans;
    }
};